src/polycubed/src/utils/netlink.cpp

```diff
@@ -616,5 +632,212 @@ void Netlink::detach_from_xdp(const std::string &iface, int attach_flags) {
+  };
+  int nlfd;
+
+  memset(&nladdr, 0, sizeof(struct sockaddr_nl));
+  nladdr.nl_family = AF_NETLINK;
+  nladdr.nl_pid = 0;
+  nladdr.nl_groups = 0;
+
+  nlfd = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+  if (nlfd < 0) {
+    free(nlmsg);
+    logger->error("netlink_nl_send: Unable to open socket");
+  }
+
+  if (sendmsg(nlfd, &msg, 0) < 0) {
+    free(nlmsg);
+    close(nlfd);
+    logger->error("netlink_nl_send: Unable to sendmsg");
+  }
+
+  /* read the reply message to check that there are no errors */
+  if (recvmsg(nlfd, &msg, 0) < 0) {
+    free(nlmsg);
+    close(nlfd);
+    logger->error("netlink_nl_send: Unable to recvmsg");
+  }
+
+  /* if there is an error return error code */
+  if (nlmsg->nlmsg_type == NLMSG_ERROR) {
+    struct nlmsgerr *err = (struct nlmsgerr*)NLMSG_DATA(nlmsg);
+    free(nlmsg);
+    close(nlfd);
+    return err->error;
+  }
+
+  free(nlmsg);
+  close(nlfd);
+  return 0;
+}
+
+void Netlink::set_iface_status(const std::string &iface, STATUS_IFACE status) {
+  struct nlmsghdr *nlmsg = netlink_alloc();
+  struct nl_req *unr = (struct nl_req *)nlmsg;
+
+  int index = get_iface_index(iface);
+  if (index == -1) {
+    logger->error("set_iface_status: iface {0} does not exist", iface);
+    return;
+  }
+
+  unr->ifinfomsg.ifi_index = index;
+  unr->ifinfomsg.ifi_change |= IFF_UP;
+  if (status == STATUS_IFACE::UP)
+    unr->ifinfomsg.ifi_flags |= IFF_UP;
+  else if (status == STATUS_IFACE::DOWN)
+    unr->ifinfomsg.ifi_flags |= ~IFF_UP;
+
+  netlink_nl_send(nlmsg);
+}
+
+void Netlink::set_iface_mac(const std::string &iface, const std::string &mac) {
+  struct rtnl_link *link;
+  struct rtnl_link *old_link;
+  struct nl_sock *sk;
+  struct nl_cache *link_cache;
+
+  sk = nl_socket_alloc();
+  if (nl_connect(sk, NETLINK_ROUTE) < 0) {
+    logger->error("set_iface_mac: Unable to open socket");
+    return;
+  }
+
+  link = rtnl_link_alloc();
+  if (!link) {
+    nl_close(sk);
+    logger->error("set_iface_mac: Invalid link");
+    return;
+  }
+
+  rtnl_link_set_name(link, iface.c_str());
+
+  struct nl_addr* addr;
+  addr = nl_addr_build(AF_LLC, ether_aton(mac.c_str()), ETH_ALEN);
+  rtnl_link_set_addr(link, addr);
+
+  if (rtnl_link_alloc_cache(sk, AF_UNSPEC, &link_cache) < 0) {
+    nl_close(sk);
+    logger->error("set_iface_mac: Unable to allocate cache");
+    return;
+  }
+
+  old_link = rtnl_link_get_by_name(link_cache, iface.c_str());
+  if (rtnl_link_change(sk, old_link, link, 0) < 0) {
+    nl_close(sk);
+    logger->error("set_iface_mac: Unable to change link");
+    return;
+  }
+
+  rtnl_link_put(link);
+  nl_close(sk);
+}
+
+void Netlink::set_iface_ip(const std::string &iface, const std::string &ip, int prefix) {
+  struct nlmsghdr *nlmsg = netlink_ip_alloc();
+  struct nl_ipreq *uni = (struct nl_ipreq *)nlmsg;
+  struct rtattr *rta;
+  struct in_addr ia;
+
+  int index = get_iface_index(iface);
+  if (index == -1) {
+    logger->error("set_iface_ip: iface {0} does not exist", iface);
```

Comment:
How can the caller know if there was an error? 
It is better to throw here, the caller can log the error if he wants.